cpmpy/tools/explain/mus.py

```diff
@@ -108,6 +111,55 @@ def do_recursion(soft, hard, delta):
     return [dmap[a] for a in core]
 
 
+def optimal_mus(soft, hard=[], weights=1, solver="ortools", hs_solver="ortools"):
+    """
+        Find an optimal MUS according to a linear objective function.
+        Works by iteratively generating correction subsets and computing optimal hitting sets to those enumerated sets.
+        For better performance of the algorithm, use an incemental solver to compute the hitting sets such as Gurobi.
+
+        Assumption-based implementation for solvers that support s.solve(assumptions=...)
+        More naive version available as `optimal_mus_naive` to use with other solvers.
+
+        CPMpy implementation loosely based on the "SMUS" algorithm from
+            Ignatiev, Alexey, et al. "Smallest MUS extraction with minimal hitting set dualization."
+            International Conference on Principles and Practice of Constraint Programming. Cham: Springer International Publishing, 2015.
+        and OCUS from:
+            Gamba, Emilio, Bart Bogaerts, and Tias Guns. "Efficiently explaining CSPs with unsatisfiable subset optimization."
+            Journal of Artificial Intelligence Research 78 (2023): 709-746.
+
+    """
+
+
+    model, soft, assump = make_assump_model(soft, hard)
+    dmap = dict(zip(assump, soft))
+
+    s = cp.SolverLookup.get(solver, model)
+    # s.solution_hint(assump, [1]*len(assump)) # causes weirdness in OR-Tools: https://github.com/google/or-tools/issues/4324
+    assert s.solve(assumptions=assump) is False
+
+    # initialize hitting set solver
+    if is_num(weights):
+        weights = np.ones(len(assump), dtype=int)
+    hs_solver = cp.SolverLookup.get(hs_solver)
+    hs_solver.minimize(cp.sum(assump * np.array(weights)))
+
+    while hs_solver.solve() is True:
+        hitting_set = [a for a in assump if a.value()]
+        if s.solve(assumptions=hitting_set) is False:
+            break
+
+        # SAT, use hitting set as start for disjoint MCS enumeration
+        sat_subset = list(hitting_set)  # make new list
+        while s.solve(assumptions=sat_subset) is True:
```

Comment:
you are calling solve again with the same assumptions in the first iteration...
I think you should first add the 'just found' MCS, and then comment that you will also search for and extend with disjoint ones...
If you dont like the code duplication (this implementation is exceptionally clean currently...), you can also make it a while True:, with the solve call at the end and from which you break if False?

I guess I would favor something like:
mcs = [...]
hs_solver +=

if add_disjoint_mcs:
  sat_subset = hs + mcs...              <--- maybe a comment of why this works ; )
  while s.solve() is True:
   ...
   (

Reply:
Refactored this to make it two clear steps.